### ISSUE TITLE
[C++] Fix and re-enable Asof Join Backpresure test flakiness

### ISSUE THREAD

**Issue by icexelloss** [apache/arrow#36248]
### Describe the enhancement requested

See https://github.com/apache/arrow/pull/36020#issuecomment-1602538003

The backpressure test in asof join is flaky. We should fix it.

Update (20230707):

In #36331 we decided to temporarily disable the flaky asof join backpressure test. So this jira is meant to track the work to fix the test and reenable it.

### Component(s)

C++

**Comment by rtpsw** [comment]
@bkietz, do you have any more info about this issue? In which CI jobs did you observe the test fail? Did you always observe the same assertion fail? Did you notice how frequently the failure happens?

**Comment by bkietz** [comment]
So far I've only seen it once in [this run of AMD64 Ubuntu 22.04 C++ ASAN UBSAN](https://github.com/apache/arrow/actions/runs/5340154873/jobs/9679635756?pr=36020#step:6:2882). I don't have a lot of information, sorry.

For trying to diagnose issues like this it'd be neat to have a label or tag or something which can be grepped for across all PRs. @amol- what if we had something like `@github-actions register-flaky $link_to_CI_run`, which saves the whole log to longer-term storage? That'd both make it more formal/obvious when you ignore a flaky failure in a PR and would make it possible for devs hunting those failures to find them

**Comment by raulcd** [comment]
@icexelloss is this a duplicate for: https://github.com/apache/arrow/issues/36331
Should this be closed?

**Comment by icexelloss** [comment]
@raulcd Not really. #36331 merely disabled the test. This issue is meant to fixing/reenable the test. Let me update the description here.

**Comment by raulcd** [comment]
ok, cool! @icexelloss but is not a release blocker, right? I can move the milestone to 14.0.0?

**Comment by icexelloss** [comment]
Correct not a release blocker.

**Comment by thisisnic** [comment]
This issue has been marked as stale because it has had no activity in the past 2 years. Please remove the stale label or comment below, or this issue will be closed in 14 days.
